web-api/src/personalization/knn.rs

```diff
@@ -23,14 +23,14 @@ use xayn_ai_coi::{compute_coi_weights, PositiveCoi};
 use crate::{
     error::common::InternalError,
     models::{DocumentId, PersonalizedDocument},
-    storage::{self, KnnSearchParams},
+    storage::{self, KnnSearchParams, SearchStrategy},
     Error,
 };
 
 /// KNN search based on Centers of Interest.
-pub(super) struct CoiSearch<'a, I, J> {
+pub(super) struct CoiSearch<'a, I> {
     pub(super) interests: I,
-    pub(super) excluded: J,
+    pub(super) excluded: &'a [DocumentId],
```

Comment:
❓ why make it less flexible? for example now we can't use iterators on hashsets anymore without reallocation.

Reply:
1. Because only then can we attach some of the method you wanted to have attached to `KnnSearchParams` to it without involving additional clones of which you can't be sure to be cheap in the place they are done.

2. Because we don't use the generics but they add a bunch of complexity all through the code.

Reply:
>  reduction of complexity across the code could also be done via a type alias.

Don't work for most of the complexity introduced here as it involves many `impl Trait`'s and complex where bounds.

I mean we went from `impl IntoIterator<Item=&'a DocumentId>` to `impl IntoIterator<IntoIter = impl ExactSizeIterator<Item = &'a DocumentId> + Clone>` .

For a situation where in all places in the code we use only things which we for "free" (deref,ref) can turn into a `&[DocumentId]` this is rather a lot of complexity in a situation where it not even clear that if we maybe use some other input the conversion to `&[DocumentId]` is more expensive then the cost of additional collects in lower level code.